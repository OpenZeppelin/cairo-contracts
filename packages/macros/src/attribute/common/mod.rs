pub mod text_span;
