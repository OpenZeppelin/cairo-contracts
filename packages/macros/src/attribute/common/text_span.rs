use std::collections::{HashMap, VecDeque};

use cairo_lang_macro::{quote, TextSpan, Token, TokenStream, TokenTree};
use cairo_lang_parser::utils::SimpleParserDatabase;
use cairo_lang_syntax::node::with_db::SyntaxNodeWithDb;

fn make_ident(text: &str, span: TextSpan) -> TokenTree {
    TokenTree::Ident(Token::new(text, span))
}

fn tokenize_str(s: &str, db: &SimpleParserDatabase) -> TokenStream {
    let syntax_node = db.parse_virtual(s).unwrap();
    let syntax_node_with_db = SyntaxNodeWithDb::new(&syntax_node, db);
    quote! {#syntax_node_with_db}
}

/// Merge spans from `initial_tokens` into a tokenized version of `final_output`.
/// - If a token text in `final_output` matches one from `initial_tokens` (by exact text),
///   it inherits that tokenâ€™s span (first-come, first-served).
/// - Otherwise, it gets `TextSpan::call_site()`.
pub fn merge_spans_from_initial(
    initial_tokens: &[TokenTree],
    final_output: &str,
    db: &SimpleParserDatabase,
) -> Vec<TokenTree> {
    // Build a multimap: text -> queue of spans (to support duplicates, left-to-right)
    let mut span_index: HashMap<String, VecDeque<TextSpan>> = HashMap::new();
    for tt in initial_tokens {
        let TokenTree::Ident(tok) = tt;
        let text = tok.content.to_string();
        span_index
            .entry(text)
            .or_default()
            .push_back(tok.span.clone());
    }

    // Tokenize final output
    let final_tokens = tokenize_str(final_output, db)
        .tokens
        .iter()
        .map(|tt| {
            let TokenTree::Ident(tok) = tt;
            tok.content.to_string()
        })
        .collect::<Vec<_>>();

    // Rebuild final tokens, reusing spans when the text matches; else call_site.
    let mut out = Vec::with_capacity(final_tokens.len());
    for text in final_tokens {
        if let Some(queue) = span_index.get_mut(&text) {
            if let Some(span) = queue.pop_front() {
                out.push(make_ident(&text, span));
                if queue.is_empty() {
                    // keep the map clean
                    span_index.remove(&text);
                }
                continue;
            }
        }
        out.push(make_ident(&text, TextSpan::call_site()));
    }

    out
}
