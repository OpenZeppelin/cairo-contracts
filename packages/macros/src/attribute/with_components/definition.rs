use cairo_lang_formatter::format_string;
use cairo_lang_macro::{attribute_macro, Diagnostic, ProcMacroResult, TokenStream};
use cairo_lang_parser::utils::SimpleParserDatabase;
use regex::Regex;

use super::{components::AllowedComponents, parser::WithComponentsParser};

/// Inserts multiple component dependencies into a modules codebase.
#[attribute_macro]
pub fn with_components(attribute_stream: TokenStream, item_stream: TokenStream) -> ProcMacroResult {
    let args = parse_args(&attribute_stream.to_string());

    // 1. Get the components info (if valid)
    let mut components_info = vec![];
    let empty_result = ProcMacroResult::new(TokenStream::empty());
    for arg in args {
        let maybe_component = AllowedComponents::from_str(&arg);
        match maybe_component {
            Ok(component) => {
                components_info.push(component.get_info());
            }
            Err(err) => {
                return empty_result.with_diagnostics(err.into());
            }
        }
    }

    // 2. Parse the item stream
    let db = SimpleParserDatabase::default();
    let (content, diagnostics) = match db.parse_virtual(item_stream.to_string()) {
        Ok(node) => WithComponentsParser::new(node, &components_info).parse(&db),
        Err(err) => {
            let error = Diagnostic::error(err.format(&db));
            return empty_result.with_diagnostics(error.into());
        }
    };

    let formatted_content = if !content.is_empty() {
        format_string(&db, content)
    } else {
        content
    };

    ProcMacroResult::new(TokenStream::new(formatted_content)).with_diagnostics(diagnostics)
}

/// Parses the arguments from the attribute stream.
fn parse_args(text: &str) -> Vec<String> {
    let re = Regex::new(r"(\w+)").unwrap();
    let matches = re.find_iter(text);
    matches.map(|m| m.as_str().to_string()).collect()
}

#[cfg(test)]
mod tests {
    use super::parse_args;

    #[test]
    fn test_parse_args() {
        let attribute = "(ERC20, Ownable)";
        let result = parse_args(attribute);
        assert_eq!(result, vec!["ERC20", "Ownable"]);

        let attribute = "ERC20";
        let result = parse_args(attribute);
        assert_eq!(result, vec!["ERC20"]);

        let attribute = "(Ownable, ERC20, Other, Another)";
        let result = parse_args(attribute);
        assert_eq!(result, vec!["Ownable", "ERC20", "Other", "Another"]);
    }
}
