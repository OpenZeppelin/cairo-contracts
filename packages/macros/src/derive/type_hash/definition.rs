use cairo_lang_formatter::format_string;
use cairo_lang_macro::{attribute_macro, Diagnostic, Diagnostics, ProcMacroResult, TokenStream};
use cairo_lang_parser::utils::SimpleParserDatabase;
use cairo_lang_plugins::plugins::utils::PluginTypeInfo;
use cairo_lang_starknet_classes::keccak::starknet_keccak;
use cairo_lang_syntax::node::{ast, TypedSyntaxNode};
use cairo_lang_syntax::node::{db::SyntaxGroup, SyntaxNode};
use convert_case::{Case, Casing};
use indoc::formatdoc;
use regex::Regex;

use crate::type_hash::parser::TypeHashParser;

use super::diagnostics::errors;
use super::parser::parse_string_arg;

/// Derive macro that generates a SNIP-12 type hash constant for a struct.
///
/// Example:
/// ```
/// #[type_hash]
/// pub struct MyStruct {
///     pub some_member: felt252,
/// }
///
/// // Generates:
/// pub const MY_STRUCT_TYPE_HASH: felt252 = 0x[HASH];
/// ```
#[attribute_macro]
pub fn type_hash(attr_stream: TokenStream, item_stream: TokenStream) -> ProcMacroResult {
    // 1. Parse the attribute stream
    let config = match parse_args(&attr_stream.to_string()) {
        Ok(config) => config,
        Err(err) => {
            return ProcMacroResult::new(TokenStream::empty()).with_diagnostics(err.into());
        }
    };

    // 2. Parse the item stream
    let db = SimpleParserDatabase::default();
    let content = match db.parse_virtual(item_stream.to_string()) {
        Ok(node) => handle_node(&db, node, &config),
        Err(err) => {
            let error = Diagnostic::error(err.format(&db));
            return ProcMacroResult::new(TokenStream::empty()).with_diagnostics(error.into());
        }
    };

    // 3. Format the expanded content
    let (formatted_content, diagnostics) = match content {
        Ok(content) => (format_string(&db, content), Diagnostics::new(vec![])),
        Err(err) => (String::new(), err.into()),
    };

    // 4. Return the result
    let result = TokenStream::new(item_stream.to_string() + &formatted_content);
    ProcMacroResult::new(result).with_diagnostics(diagnostics)
}

/// This attribute macro is used to specify an override for the SNIP-12 type.
///
/// It doesn't modify the source code directly, but it is used in the type hash parser to generate the new type hash.
///
/// Example:
/// ```
/// #[type_hash]
/// pub struct MyStruct {
///     #[snip12(name: "Some Member", kind: "shortstring")]
///     pub some_member: felt252,
/// }
/// ```
#[attribute_macro]
pub fn snip12(attr_stream: TokenStream, item_stream: TokenStream) -> ProcMacroResult {
    // Validate the received format
    let re1 = Regex::new(r"^\(name: (.*), kind: (.*)\)$").unwrap();
    let re2 = Regex::new(r"^\(kind: (.*), name: (.*)\)$").unwrap();
    let re3 = Regex::new(r"^\(kind: (.*)\)$").unwrap();
    let re4 = Regex::new(r"^\(name: (.*)\)$").unwrap();
    let s = attr_stream.to_string();

    if re1.is_match(&s) || re2.is_match(&s) || re3.is_match(&s) || re4.is_match(&s) {
        ProcMacroResult::new(item_stream)
    } else {
        let error = Diagnostic::error(errors::INVALID_SNIP12_ATTRIBUTE_FORMAT);
        ProcMacroResult::new(item_stream).with_diagnostics(error.into())
    }
}

/// Configuration for the type hash attribute.
///
/// Represents the arguments passed to the type_hash attribute.
///
/// Example:
/// ```
/// #[type_hash(name: "MyStruct", debug: true)]
/// ```
pub struct TypeHashArgs {
    pub name: String,
    pub debug: bool,
}

/// Parses the arguments passed to the type_hash attribute and
/// returns a TypeHashArgs struct containing the parsed arguments.
fn parse_args(s: &str) -> Result<TypeHashArgs, Diagnostic> {
    let mut args = TypeHashArgs {
        name: String::new(),
        debug: false,
    };
    // If the attribute is empty, return the default config
    if s.is_empty() || s == "()" {
        return Ok(args);
    }

    let allowed_args = ["name", "debug"];
    let allowed_args_re = allowed_args.join("|");

    let re = Regex::new(&format!(
        r#"^\(({}): ([\w"' ])+(?:, ({}): ([\w"' ])+)*\)$"#,
        allowed_args_re, allowed_args_re
    ))
    .unwrap();

    if re.is_match(s) {
        // Remove the parentheses
        let s = &s[1..s.len() - 1];
        for arg in s.split(",") {
            let parts = arg.split(":").collect::<Vec<&str>>();
            let name = parts[0].trim();
            let value = parts[1].trim();
            match name {
                "name" => args.name = parse_string_arg(value)?,
                "debug" => args.debug = value == "true",
                // This should be unreachable as long as the regex is correct
                _ => panic!("Invalid argument: {}", name),
            }
        }
        Ok(args)
    } else {
        Err(Diagnostic::error(
            errors::INVALID_TYPE_HASH_ATTRIBUTE_FORMAT,
        ))
    }
}

fn handle_node(
    db: &dyn SyntaxGroup,
    node: SyntaxNode,
    args: &TypeHashArgs,
) -> Result<String, Diagnostic> {
    let typed = ast::SyntaxFile::from_syntax_node(db, node);
    let items = typed.items(db).elements(db);

    let Some(item_ast) = items.first() else {
        let error = Diagnostic::error(errors::EMPTY_TYPE_FOUND);
        return Err(error);
    };

    // Generate type hash for structs/enums only
    match item_ast {
        ast::ModuleItem::Struct(_) | ast::ModuleItem::Enum(_) => {
            // It is safe to unwrap here because we know the item is a struct
            let plugin_type_info = PluginTypeInfo::new(db, item_ast).unwrap();
            generate_code(db, &plugin_type_info, args)
        }
        _ => {
            let error = Diagnostic::error(errors::NOT_VALID_TYPE_TO_DECORATE);
            Err(error)
        }
    }
}

/// Generates the code for the type hash constant.
fn generate_code(
    db: &dyn SyntaxGroup,
    plugin_type_info: &PluginTypeInfo,
    args: &TypeHashArgs,
) -> Result<String, Diagnostic> {
    let mut parser = TypeHashParser::new(plugin_type_info);
    let type_hash_string = parser.parse(db, args)?;
    let type_hash = starknet_keccak(type_hash_string.as_bytes());
    let type_name = plugin_type_info.name.as_str().to_case(Case::UpperSnake);

    let debug_string = if args.debug {
        formatdoc!(
            r#"
            pub fn __{type_name}_encoded_type() {{
                println!("{}");
            }}"#,
            type_hash_string.replace("\"", "\\\"")
        )
    } else {
        String::new()
    };

    let code = formatdoc!(
        "
        {debug_string}
        pub const {type_name}_TYPE_HASH: felt252 = 0x{:x};

        ",
        type_hash
    );

    Ok(code)
}
