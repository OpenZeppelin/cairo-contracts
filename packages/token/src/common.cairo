pub mod erc2981;
