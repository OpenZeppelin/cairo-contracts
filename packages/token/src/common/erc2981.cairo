pub mod erc2981;
pub mod interface;

pub use erc2981::{ERC2981Component, DefaultConfig};
pub use interface::{IERC2981Dispatcher, IERC2981DispatcherTrait};
