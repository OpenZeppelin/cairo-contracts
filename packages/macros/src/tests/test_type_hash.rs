use cairo_lang_macro::TokenStream;
use indoc::indoc;
use insta::assert_snapshot;

use crate::type_hash::definition::type_hash_jgjpoopqerqnq as type_hash;

use super::common::format_proc_macro_result;

#[test]
fn test_empty_input() {
    let item = indoc!(
        "
        "
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_empty_struct() {
    let item = indoc!(
        "
        pub struct MyType {}
        "
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_empty_enum() {
    let item = indoc!(
        "
        pub enum MyEnum {}
        "
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_snip12_attribute_empty() {
    let item = indoc!(
        "
        pub struct MyType {
            #[snip12()]
            pub name: felt252,
        }
        "
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_basic_types() {
    // Basic types list:
    // - Felt
    // - ShortString
    // - ClassHash
    // - ContractAddress
    // - Timestamp
    // - Selector
    // - U128
    // - I128
    let item = indoc!(
        r#"
        pub struct MyType {
            pub name: felt252,
            #[snip12(kind: "shortstring")]
            pub version: felt252,
            pub class_hash: ClassHash,
            pub contract_address: ContractAddress,
            #[snip12(kind: "timestamp")]
            pub timestamp: u128,
            #[snip12(kind: "selector")]
            pub selector: felt252,
            pub u128_member: u128,
            pub i128_member: i128,
        }
        "#
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_basic_types_enum() {
    let item = indoc!(
        r#"
        pub enum MyEnum {
            Variant1: felt252,
            Variant2: ClassHash,
            Variant3: ContractAddress,
            Variant4: u128,
            Variant5: i128,
            #[snip12(kind: "shortstring")]
            Variant6: felt252,
            #[snip12(kind: "timestamp")]
            Variant7: u128,
            #[snip12(kind: "selector")]
            Variant8: felt252,
        }
        "#
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_preset_types() {
    // Preset types list:
    // - TokenAmount
    // - NftId
    // - U256
    let item = indoc!(
        "
        pub struct MyType {
            pub token_amount: TokenAmount,
            pub nft_id: NftId,
            pub u256: u256,
        }
        "
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_preset_types_enum() {
    let item = indoc!(
        "
        pub enum MyEnum {
            Variant1: TokenAmount,
            Variant2: NftId,
            Variant3: u256,
        }
        "
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}
#[test]
fn test_with_inner_starknet_domain() {
    let item = indoc!(
        "
        pub struct MyType {
            pub starknet_domain: StarknetDomain,
        }
        "
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_with_inner_u256_type() {
    let item = indoc!(
        "
        pub struct MyType {
            // TokenAmount type contains u256, which should be resolved
            // and appended to the final type hash.
            pub token_amount: TokenAmount,
        }
        "
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_with_inner_u256_type_enum() {
    let item = indoc!(
        "
        pub enum MyEnum {
            // TokenAmount type contains u256, which should be resolved
            // and appended to the final type hash.
            Variant1: TokenAmount,
        }
        "
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_potential_duplicate_types() {
    let item = indoc!(
        "
        pub struct MyType {
            // TokenAmount type contains u256, which should be resolved
            // and appended to the final type hash.
            pub token_amount: TokenAmount,
            pub token_amount_2: TokenAmount,
            pub number: u256,
        }
        "
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_potential_duplicate_types_enum() {
    let item = indoc!(
        "
        pub enum MyEnum {
            // TokenAmount type contains u256, which should be resolved
            // and appended to the final type hash.
            Variant1: TokenAmount,
            Variant2: TokenAmount,
            Variant3: u256,
        }
        "
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_complex_struct_type() {
    let item = indoc!(
        r#"
        pub struct MyType {
            pub token_amount: TokenAmount,
            pub token_amount_2: TokenAmount,
            pub number: u256,
            #[snip12(kind: "shortstring")]
            pub version: felt252,
            pub class_hash: ClassHash,
            pub contract_address: ContractAddress,
            #[snip12(kind: "timestamp")]
            pub timestamp: u128,
            #[snip12(kind: "selector")]
            pub selector: felt252,
            pub u128_member: u128,
            pub i128_member: i128,
        }
        "#
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_complex_enum_type() {
    let item = indoc!(
        r#"
        pub enum MyEnum {
            Variant1: TokenAmount,
            Variant2: TokenAmount,
            Variant3: u256,
            #[snip12(kind: "shortstring")]
            Variant4: felt252,
            Variant5: ClassHash,
            Variant6: ContractAddress,
            #[snip12(kind: "timestamp")]
            Variant7: u128,
            #[snip12(kind: "selector")]
            Variant8: felt252,
            Variant9: u128,
            Variant10: i128,
        }
        "#
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_with_inner_custom_type() {
    let item = indoc!(
        "
        pub struct MyType {
            pub name: felt252,
            pub version: felt252,
            pub chain_id: felt252,
            pub revision: felt252,
            pub member: InnerCustomType,
        }
        "
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_with_tuple() {
    let item = indoc!(
        "
        pub struct MyType {
            pub member: (felt252, felt252, ClassHash, NftId),
        }
        "
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_with_tuple_enum() {
    let item = indoc!(
        "
        pub enum MyEnum {
            Variant1: (felt252, felt252, ClassHash, NftId),
            Variant2: TokenAmount,
            Variant3: (ContractAddress,),
        }
        "
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_with_empty_tuple() {
    // This doesn't make sense, but it's a valid type.
    let item = indoc!(
        "
        pub struct MyType {
            pub member: (),
        }
        "
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_with_empty_tuple_enum() {
    let item = indoc!(
        "
        pub enum MyEnum {
            Variant1: TokenAmount,
            Variant2: (),
        }
        "
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_with_array() {
    let item = indoc!(
        "
        pub struct MyType {
            pub member1: Array<felt252>,
            pub member2: Array<TokenAmount>,
            pub member3: Array<ClassHash>,
            pub member4: Array<ContractAddress>,
            pub member5: Array<u128>,
            pub member6: Array<i128>,
        }
        "
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_with_span() {
    let item = indoc!(
        "
        pub struct MyType {
            pub member1: Span<felt252>,
            pub member2: Span<TokenAmount>,
            pub member3: Span<ClassHash>,
            pub member4: Span<ContractAddress>,
            pub member5: Span<u128>,
            pub member6: Span<i128>,
        }
        "
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_with_tuple_and_attribute() {
    let item = indoc!(
        r#"
        pub struct MyType {
            #[snip12(kind: "(shortstring, felt252, ClassHash, NftId)")]
            pub member: (felt252, felt252, ClassHash, NftId),
        }
        "#
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_starknet_domain() {
    let item = indoc!(
        r#"
        pub struct StarknetDomain {
            #[snip12(kind: "shortstring")]
            pub name: felt252,
            #[snip12(kind: "shortstring")]
            pub version: felt252,
            #[snip12(kind: "shortstring")]
            pub chainId: felt252,
            #[snip12(kind: "shortstring")]
            pub revision: felt252,
        }
        "#
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_complex_struct_with_collection_types() {
    let item = indoc!(
        r#"
        pub struct MyType {
            pub member1: (felt252, felt252, ClassHash, NftId),
            pub member2: Array<TokenAmount>,
            pub member3: Span<ClassHash>,
            pub member4: (ContractAddress, TokenAmount),
            pub member5: Array<ContractAddress>,
            pub member6: (),
            #[snip12(kind: "(timestamp, shortstring)")]
            pub member7: (u128, felt252),
            pub member8: (ContractAddress,),
            pub member9: (TokenAmount, (felt252, ClassHash), NftId),
            pub member10: (Array<TokenAmount>, Array<ContractAddress>),
            pub member11: Array<(TokenAmount, ContractAddress, Array<felt252>)>,
            pub member12: Array<Array<(Array<TokenAmount>, Array<ContractAddress>, Array<felt252>)>>,
        }
        "#
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_complex_struct_with_collection_types_custom_names() {
    let item = indoc!(
        r#"
        pub struct MyType {
            #[snip12(name: "Member 1")]
            pub member1: (felt252, felt252, ClassHash, NftId),
            #[snip12(name: "Member 2")]
            pub member2: Array<TokenAmount>,
            #[snip12(name: "Member 3")]
            pub member3: Span<ClassHash>,
            #[snip12(name: "Member 4")]
            pub member4: (ContractAddress, TokenAmount),
            #[snip12(name: "Member 5")]
            pub member5: Array<ContractAddress>,
            #[snip12(name: "Member 6")]
            pub member6: (),
            #[snip12(name: "Member 7", kind: "(timestamp, shortstring)")]
            pub member7: (u128, felt252),
            #[snip12(name: "Member 8")]
            pub member8: (ContractAddress,),
            #[snip12(name: "Member 9")]
            pub member9: (TokenAmount, (felt252, ClassHash), NftId),
            #[snip12(name: "Member 10")]
            pub member10: (Array<TokenAmount>, Array<ContractAddress>),
            #[snip12(name: "Member 11")]
            pub member11: Array<(TokenAmount, ContractAddress, Array<felt252>)>,
            #[snip12(name: "Member 12")]
            pub member12: Array<Array<(Array<TokenAmount>, Array<ContractAddress>, Array<felt252>)>>,
        }
        "#
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_complex_enum_with_collection_types() {
    let item = indoc!(
        r#"
        pub enum MyEnum {
            Variant1: (felt252, felt252, ClassHash, NftId),
            Variant2: Array<TokenAmount>,
            Variant3: Span<ClassHash>,
            Variant4: (ContractAddress, TokenAmount),
            Variant5: Array<ContractAddress>,
            Variant6: (),
            #[snip12(kind: "(timestamp, shortstring)")]
            Variant7: (u128, felt252),
            Variant8: (ContractAddress,),
            Variant9: (TokenAmount, (felt252, ClassHash), NftId),
            Variant10: (Array<TokenAmount>, Array<ContractAddress>),
            Variant11: Array<(TokenAmount, ContractAddress, Array<felt252>)>,
            Variant12: Array<Array<(Array<TokenAmount>, Array<ContractAddress>, Array<felt252>)>>,
        }
        "#
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_complex_enum_with_collection_types_custom_names() {
    let item = indoc!(
        r#"
        pub enum MyEnum {
            #[snip12(name: "Variant 1")]
            Variant1: (felt252, felt252, ClassHash, NftId),
            #[snip12(name: "Variant 2")]
            Variant2: Array<TokenAmount>,
            #[snip12(name: "Variant 3")]
            Variant3: Span<ClassHash>,
            #[snip12(name: "Variant 4")]
            Variant4: (ContractAddress, TokenAmount),
            #[snip12(name: "Variant 5")]
            Variant5: Array<ContractAddress>,
            #[snip12(name: "Variant 6")]
            Variant6: (),
            #[snip12(name: "Variant 7", kind: "(timestamp, shortstring)")]
            Variant7: (u128, felt252),
            #[snip12(name: "Variant 8")]
            Variant8: (ContractAddress,),
            #[snip12(name: "Variant 9")]
            Variant9: (TokenAmount, (felt252, ClassHash), NftId),
            #[snip12(name: "Variant 10")]
            Variant10: (Array<TokenAmount>, Array<ContractAddress>),
            #[snip12(name: "Variant 11")]
            Variant11: Array<(TokenAmount, ContractAddress, Array<felt252>)>,
            #[snip12(name: "Variant 12")]
            Variant12: Array<Array<(Array<TokenAmount>, Array<ContractAddress>, Array<felt252>)>>,
        }
        "#
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_name_attribute() {
    let item = indoc!(
        r#"
        pub enum MyEnum {
            #[snip12(name: "Variant 1")]
            Variant1: (felt252, felt252, ClassHash, NftId),
            #[snip12(name: "Variant 2")]
            Variant2: Array<TokenAmount>,
            #[snip12(name: "Variant 3")]
            Variant3: Span<ClassHash>,
            #[snip12(name: "Variant 4")]
            Variant4: (ContractAddress, TokenAmount),
            #[snip12(name: "Variant 5")]
            Variant5: Array<ContractAddress>,
        }
        "#
    );
    let attr_stream = r#"(name: "My Enum")"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_debug_attribute() {
    let item = indoc!(
        "
        pub struct MyType {
            pub member: felt252,
        }
        "
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_invalid_type_hash_attribute() {
    let item = indoc!(
        "
        pub struct MyType {
            pub member: felt252,
        }
        "
    );
    let attr_stream = r#"(other: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_enum_without_explicit_variant_type() {
    let item = indoc!(
        "
        pub enum MyEnum {
            Variant1,
            Variant2,
        }
        "
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_merkletree_type() {
    let item = indoc!(
        r#"
        pub struct MyType {
            #[snip12(kind: "merkletree")]
            pub member: felt252,
        }
        "#
    );
    let attr_stream = r#"(debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

//
// Doc examples
//

#[test]
fn test_doc_example_1() {
    let item = indoc!(
        r#"
        struct MyStruct {
            #[snip12(name: "My Field")]
            my_field: felt252,
        }
        "#
    );
    let attr_stream = r#"(name: "My Struct", debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_doc_example_2() {
    let item = indoc!(
        r#"
        pub struct MyStruct {
            #[snip12(name: "Simple Felt")] // Optional custom name
            pub simple_felt: felt252,
            #[snip12(name: "Class Hash")]
            pub class_hash: ClassHash,
            #[snip12(name: "Target Token")]
            pub target: ContractAddress,
            #[snip12(name: "Timestamp", kind: "timestamp")]
            pub timestamp: u128,
            #[snip12(name: "Selector", kind: "selector")]
            pub selector: felt252,
        }
        "#
    );
    let attr_stream = r#"(name: "My Struct", debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_doc_example_3() {
    let item = indoc!(
        r#"
        pub enum MyEnum {
            #[snip12(name: "Simple Felt")]
            SimpleFelt: felt252,
            #[snip12(name: "Class Hash")]
            ClassHash: ClassHash,
            #[snip12(name: "Target Token")]
            ContractAddress: ContractAddress,
            #[snip12(name: "Timestamp", kind: "timestamp")]
            Timestamp: u128,
            #[snip12(name: "Selector", kind: "selector")]
            Selector: felt252,
        }
        "#
    );
    let attr_stream = r#"(name: "My Enum", debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_doc_example_4() {
    let item = indoc!(
        r#"
        pub struct MyStruct {
            #[snip12(name: "Member 1")]
            pub member1: Array<felt252>,
            #[snip12(name: "Member 2")]
            pub member2: Span<u128>,
            #[snip12(name: "Timestamps", kind: "Array<timestamp>")]
            pub timestamps: Array<u128>,
        }
        "#
    );
    let attr_stream = r#"(name: "My Struct", debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_doc_example_5() {
    let item = indoc!(
        r#"
        pub enum MyEnum {
            #[snip12(name: "Member 1")]
            Member1: Array<felt252>,
            #[snip12(name: "Member 2")]
            Member2: Span<u128>,
            #[snip12(name: "Timestamps", kind: "Array<timestamp>")]
            Timestamps: Array<u128>,
            #[snip12(name: "Name and Last Name", kind: "(shortstring, shortstring)")]
            NameAndLastName: (felt252, felt252),
        }
        "#
    );
    let attr_stream = r#"(name: "My Enum", debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_doc_example_6() {
    let item = indoc!(
        r#"
        pub struct MyStruct {
            #[snip12(name: "Token Amount")]
            pub token_amount: TokenAmount,
            #[snip12(name: "NFT ID")]
            pub nft_id: NftId,
            #[snip12(name: "Number")]
            pub number: u256,
        }
        "#
    );
    let attr_stream = r#"(name: "My Struct", debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

#[test]
fn test_doc_example_7() {
    let item = indoc!(
        r#"
        pub enum MyEnum {
            #[snip12(name: "Token Amount")]
            TokenAmount: TokenAmount,
            #[snip12(name: "NFT ID")]
            NftId: NftId,
            #[snip12(name: "Number")]
            Number: u256,
        }
        "#
    );
    let attr_stream = r#"(name: "My Enum", debug: true)"#;
    let result = get_string_result(item, attr_stream);
    assert_snapshot!(result);
}

fn get_string_result(item: &str, attr_stream: &str) -> String {
    let attr_stream = TokenStream::new(attr_stream.to_string());
    let item = TokenStream::new(item.to_string());
    let raw_result = type_hash(attr_stream, item);
    format_proc_macro_result(raw_result)
}
