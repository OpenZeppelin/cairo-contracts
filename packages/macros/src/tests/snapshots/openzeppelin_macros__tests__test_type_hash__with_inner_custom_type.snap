---
source: src/tests/test_type_hash.rs
expression: result
snapshot_kind: text
---
TokenStream:

pub struct MyType {
    pub name: felt252,
    pub version: felt252,
    pub chain_id: felt252,
    pub revision: felt252,
    pub member: InnerCustomType,
}


Diagnostics:

====
Error: Inner custom types are not supported yet.
====

AuxData:

None
